### DISCUSSION TITLE
Distributed Halide

### DISCUSSION THREAD

**Discussion by BachiLi** [halide/Halide#6004] [Show and tell]
Hi!

@Infinoid and @cimes-isi and me have been working on bringing back the distributed Halide code (https://github.com/shoaibkamil/distributed-halide) up to date with the current master branch. The current branch is here https://github.com/BachiLi/Halide/tree/distributed
It's still a bit premature and doesn't involve too much of the communication code, but the basic infrastructure is there.

We want to initiate the discussion about the plan for merging the distributed Halide code to the master. How do people think this should happen? What would be a good first PR? How do we handle MPI build/dependency? How do we test it?

Tagging people who might want to join the conversation: @jrk @abadams @shoaibkamil @alexreinking @steven-johnson @zvookin 

Happy to chat in person as well.

**Comment by jrk** [comment]
One note: this currently doesn’t support any communication across a distributed dimension, correct?

**Reply by BachiLi** [reply]
Yes. Shouldn't be that hard to add though.

**Comment by alexreinking** [comment]
Definitely something for the dev meeting. I think the biggest roadblock by far will be getting CI set up for this; whoever wants to resurrect this code should be responsible for maintaining it, including taking the lead writing the CI scripts.

**Reply by Infinoid** [reply]
> I think the biggest roadblock by far will be getting CI set up for this; whoever wants to resurrect this code should be responsible for maintaining it, including taking the lead writing the CI scripts.

I think that for basic functionality, just running two processes within the normal github-hosted runner should be fine.  Correctness tests don't care about performance.

The `CMakeLists.txt` already knows how to run the distributed tests with `mpirun -n 2`.  There will be some maximum scale to that approach, of course, but I think it can work.

**Comment by cimes-isi** [comment]
We're writing a paper about an HPC application that uses this distributed Halide implementation.  I think Halide is well-positioned for some types of HPC applications, and we are still interested in getting the distributed support merged.  It would likely enable further research in using Halide for HPC.

I'm happy to share the paper draft privately with folks if they're interested.  Cheers.
